File: algorithms/07_typescript/sorting/selection_sort.ts
export function selectionSort(arr: number[]): number[] {
  const a = arr.slice();
  for (let i = 0; i < a.length - 1; i++) {
    let minIdx = i;
    for (let j = i + 1; j < a.length; j++) if (a[j] < a[minIdx]) minIdx = j;
    if (minIdx !== i) [a[i], a[minIdx]] = [a[minIdx], a[i]];
  }
  return a;
}

const sorted = selectionSort([64, 25, 12, 22, 11]);
if (sorted.join(",") !== "11,12,22,25,64") throw new Error("Selection sort failed");
console.log("[TypeScript Selection Sort] Sorted array verified");

File: algorithms/22_haskell/sorting/SelectionSort.hs
module SelectionSort (selectionSort) where

selectionSort :: Ord a => [a] -> [a]
selectionSort [] = []
selectionSort xs =
  let m = minimum xs
  in m : selectionSort (removeOne m xs)
  where
    removeOne _ [] = []
    removeOne y (z:zs)
      | y == z = zs
      | otherwise = z : removeOne y zs

main :: IO ()
main = do
  let data' = [33, 7, 91, 12, 5, 5, 78, 2, 44, 19] :: [Int]
      sorted = selectionSort data'
  if sorted == [2, 5, 5, 7, 12, 19, 33, 44, 78, 91]
    then putStrLn ("[Haskell SelectionSort] Selection sort verified: " ++ show sorted)
    else error "SelectionSort verification failed"
